### ISSUE TITLE
Focus on migrating from mxgraph-type-definitions

### ISSUE THREAD

**Issue by hungtcs** [typed-mxgraph/typed-mxgraph#20]
Any issues due to migration can be discussed and dealt with here. @tbouffard

**Comment by ViviWilson** [comment]
Hello,
I'm migrating from mxgraph-type-definitions. I have no errors in Visual Studio Code and it builds without a problem, but when I running the application I get an error saying that mxRectangleShape is not defined.
`ERROR Error: Uncaught (in promise): TypeError: class heritage mxgraph__WEBPACK_IMPORTED_MODULE_0__.mxRectangleShape is not an object or null
./src/app/core/mxgraph/shapes/node-shapes.ts@http://localhost:4200`

Thanks for your help.

**Comment by tbouffard** [comment]
@KonstantinWilson can you provide an extract of your code using `mxRectangleShape` and the list of imports related to mxgraph?
I had the same kind of issues when trying to extend mxgraph classes based on https://github.com/typed-mxgraph/typed-mxgraph-demo in the past. They were due to wrong imports or wrong usage of objects returned by the mxgraph factory function.

**Comment by ViviWilson** [comment]
@tbouffard
Thanks for your response.
```
import { mxActor, mxRectangle, mxRectangleShape } from "mxgraph";

export class InitializationShape extends mxRectangleShape {
    constructor(
        bounds: mxRectangle,
        fill: string,
        stroke: string,
        strokewidth?: number
    ) {
        super(bounds, fill, stroke, strokewidth);
    }
}
[...]
```

I also tried using the factory (from the readme) in a component
import factory from 'mxgraph';
```
import factory from 'mxgraph';
[...]

    public ngOnInit() {
        [...]
        const mx = factory({
            mxBasePath: '.'
        });
        console.log(">>> mxGraph version <<< " + mx.mxClient.VERSION);
    }
```
and it results in this:
```
ERROR TypeError: this is undefined
    js vendor.js line 5427 > srcScript:147885
    ngOnInit shell.component.ts:78
    callHook vendor.js line 5427 > srcScript:39905
    callHooks vendor.js line 5427 > srcScript:39869
    executeInitAndCheckHooks vendor.js line 5427 > srcScript:39809
    [...]
```

**Comment by hungtcs** [comment]
Hi @KonstantinWilson! the problem looks like you directly used the import, you must call `factory()` to get the mxgraph objects instead of import directly.

```typescript
import factory from 'mxgraph';

const { mxGraph, mxClient, mxRectangleShape } = factory({...});

console.log(mxRectangleShape);  // <- use in this case 
```

you can create a helper file to load mxgraph and export the objects from factory:
```typescript
import factory from 'mxgraph';

(window as any)['mxBasePath'] = 'assets/mxgraph';

const { mxGraph, mxClient, mxRectangleShape } = factory({
  // not working see https://github.com/jgraph/mxgraph/issues/479
  mxBasePath: 'assets/mxgraph',
});

export { mxGraph, mxClient, mxRectangleShape };
```

and then in you application:
```typescript
import { mxClient, mxGraph, mxRectangleShape } from './mxgraph';
```

you can not import from `mxgraph` directly.

**Comment by hungtcs** [comment]
I have updated https://github.com/typed-mxgraph/typed-mxgraph-demo

**Comment by hungtcs** [comment]
Next I will look for a way to prohibit direct import of types.

**Comment by ViviWilson** [comment]
@hungtcs 
Thank you very much, I'm one step further. The issue I have now is that for e.g. mxGraph and mxRectangle VSC is complaining:
`'mxGraph' refers to a value, but is being used as a type here. Did you mean 'typeof mxGraph'?`

**Comment by hungtcs** [comment]
@KonstantinWilson Yes, the `mxGraph` from factory method is a value, and the type is `typeof mxGraph`
```typescript
import factory from 'mxgraph';

const { mxGraph } = factory({...});
```

**Comment by ViviWilson** [comment]
@hungtcs 
Of course:
```
import { mxRectangleShape, mxActor, mxRectangle } from '@shared/modules/mxGraphFactory/mx-graph-factory.component';

export class InitializationShape extends mxRectangleShape {
    constructor(
        bounds: **mxRectangle**,
        fill: string,
        stroke: string,
        strokewidth?: number
    ) {
        super(bounds, fill, stroke, strokewidth);
    }
}
```

Error message: 
> 'mxRectangle' refers to a value, but is being used as a type here. Did you mean 'typeof mxRectangle'?

**Comment by hungtcs** [comment]
@KonstantinWilson this is a problem. It seems that the type and value are not bound together
```ts
import { mxRectangle as MxRectangle } from 'mxgraph';
import { mxClient, mxGraph, mxRectangle } from './mxgraph';

const bounds: MxRectangle = new mxRectangle(0, 0, 0, 0);
```

**Comment by hungtcs** [comment]
> @KonstantinWilson this is a problem. It seems that the type and value are not bound together
> 
> ```ts
> import { mxRectangle as MxRectangle } from 'mxgraph';
> import { mxClient, mxGraph, mxRectangle } from './mxgraph';
> 
> const bounds: MxRectangle = new mxRectangle(0, 0, 0, 0);
> ```

@tbouffard  
Is there a solution to this problem? Because the available values are not imported directly, so type and value are separated.
I checked some similar libraries, seem to have similar problems.
- https://github.com/rabpeter/ts-mxgraph
- https://github.com/lgleim/mxgraph-typings

**Comment by tbouffard** [comment]
@hungtcs I have seen this issue with the ts-mxgraph as well and I don't know if a solution exists for this problem.
For now, I do the trick you propose or something like the following, but the idea is the same.
Your syntax seems lighter even if not perfect.

```typescript
import { mxRectangle } from 'mxgraph'; // for types
import { mxgraph } from './mxgraph'; // for instantiation. This is the whole oject returned by the mxgraph factory
...
// TS can infer the type, this is just for illustration
// However, having types is required when extending methods of mxGraph class (for argument types)
const rectangle: mxRectangle = new mxgraph.mxRectangle(...); 

```

**Comment by hungtcs** [comment]
Thanks @tbouffard, I will follow up on this issue, but there is no better solution so far. @KonstantinWilson The final solution to this problem is:
```ts
// this helper file to load mxgraph objects
import factory from 'mxgraph';

(window as any)['mxBasePath'] = 'assets/mxgraph';

export default factory({
  // not working see https://github.com/jgraph/mxgraph/issues/479
  mxBasePath: 'assets/mxgraph',
});

```
and 
```ts
import mx from './mxgraph';   // <- this import for values
import { mxRectangle } from 'mxgraph'; // <- this import for types

const bounds: mxRectangle = new mx.mxRectangle(0, 0, 0, 0);

console.log(bounds);
```

https://github.com/typed-mxgraph/typed-mxgraph-demo updated.

**Comment by ViviWilson** [comment]
Hello @hungtcs , thanks for your time.
The current solution still throws following errors for types like mxGraph and mxRectangle:

> ERROR TypeError: mxgraph__WEBPACK_IMPORTED_MODULE_2__.mxGraph is not a constructor

**Comment by hungtcs** [comment]
> Hello @hungtcs , thanks for your time.
> The current solution still throws following errors for types like mxGraph and mxRectangle:
> 
> > ERROR TypeError: mxgraph__WEBPACK_IMPORTED_MODULE_2__.mxGraph is not a constructor

@KonstantinWilson It looks like you are still using `imort { mxGraph } from 'mxgraph'` as a value. Please check your code, import from `mxgraph` as types, import from you helper created above as value, you can only use the value returned by the factory.

https://github.com/typed-mxgraph/typed-mxgraph-demo/blob/master/src/app/application.ts

**Comment by tbouffard** [comment]
@hungtcs I have been able to migrate from `mxgraph-type-definitions` in [bpmn-visualization](https://github.com/process-analytics/bpmn-visualization-js/) thanks to information provided here and the examples we have.
I didn't have issues with the types lib itself, some with the mxgraph npm package integration. This was mainly when running Jest tests and the CommonJS code generation. It has been solved by setting `"esModuleInterop": true` in the `tsconfig.json` file.
